---
title: "Deletions Cohort Analysis"
author: "Joe Shaw (CS20980)"
date: today
date-format: "DD/MM/YYYY"
format: pdf
editor: visual
fig-width: 7
fig-height: 6
---

```{r}
#| label: scripts-and-packages
#| include: FALSE

library(tidyverse)
library(here)

source(here("functions/extract_pansolid_cnv_coordinates.R"))

source(here("scripts/collate_gene_transcript_information.R"))

source(here("functions/cnv_plot_functions.R"))

source(here("scripts/del_val_load_processed_data.R"))

```

```{r}
#| label: functions
#| include: FALSE

pansolid_del_coords <- extract_pansolid_cnv_coordinates(del_val_pansolid_ngs_collated,
                                                        cnv_region) |> 
  rename(fold_change = fold_change_adjusted) |> 
  filter(!is.na(fold_change))

make_del_fine_plot <- function(labno_input, gene_input, interval_input = 1000,
                               buffer_input = 5000, ymin_input = -30,
                               ymax_input = 0) {
  
  data <- pansolid_del_coords |> 
    filter(labno == labno_input & graining == "fine")
  
  fine_plot <- make_cnv_triptych_plot(make_fold_change_cnv_plot(df = data,
                                    gene = gene_input,
                                    interval = interval_input,
                                    buffer = buffer_input,
                                    ymin = ymin_input,
                                    ymax = ymax_input,
                                    title = paste0(labno_input, 
                                               " ", gene_input,
                                               " (fine)")))
  return(fine_plot)
  
}

make_del_coarse_plot <- function(labno_input, gene_input, interval_input = 1000,
                                 buffer_input = 5000, ymin_input = -30,
                               ymax_input = 0) {
  
  data <- pansolid_del_coords |> 
    filter(labno == labno_input & graining == "coarse")
  
  coarse_plot <- make_cnv_triptych_plot(make_fold_change_cnv_plot(df = data,
                                gene = gene_input,
                                interval = interval_input,
                                buffer = buffer_input,
                                ymin = ymin_input,
                                ymax = ymax_input,
                                title = paste0(labno_input, 
                                               " ", gene_input,
                                               " (coarse)")))
  
  
  return(coarse_plot)
  
}

make_del_plot <- function(labno_input, gene_input, interval_input = 1000,
                          buffer_input = 5000, ymin_input = -30, 
                          ymax_input = 0) {
  
  
  fine_plot <- make_del_fine_plot(labno_input = labno_input,
                                  gene_input = gene_input,
                                  interval_input = interval_input,
                                  buffer_input = buffer_input,
                                  ymin_input = ymin_input,
                                  ymax_input = ymax_input)
  
  coarse_plot <- make_del_coarse_plot(labno_input = labno_input,
                                  gene_input = gene_input,
                                  interval_input = interval_input,
                                  buffer_input = buffer_input,
                                  ymin_input = ymin_input,
                                  ymax_input = ymax_input)
  
  output_plot <- coarse_plot | fine_plot
    
  return(output_plot)
  
}

```

# Introduction

```{r}

gene_coord_table <- read_csv(file = paste0(config::get("data_filepath"), 
                                      "validation/DOC6283_amplifications/",
                                      "gene_lists/",
                                      "gene_coordinates.csv"),
                        col_types = "ccccdd")

del_genes <- read_csv(file = paste0(config::get("data_filepath"),
                                   "validation/DOC6567_deletions/gene_lists/",
                                   "pansolid_deletion_gene_list.csv"))

gene_coord_table |> 
  filter(gene %in% del_genes$gene) |> 
  select(gene, chromosome) |>  view()

```



{{< pagebreak >}}

# 24018922

This sample doesn't seem to have great quality on WGS. PanSolid calls a loss of SMAD4 on both coarse and fine, and this is also detected by WGS as a loss of chromosome 18.

```{r}
#| echo: FALSE

make_del_plot("24018922", "SMAD4",
              interval_input = 10000000,
              buffer_input = 10000000,
              ymin_input = -3)
```

{{< pagebreak >}}

# 24030945

This is a QIAsymphony extracted sample with results from WGS.

## CDKN2A

WGS shows a CDKN2A/B homozygous deletion on a background of monosomy 9p. PanSolid also detects the CDKN2A/B deletions.

```{r}
#| echo: FALSE

make_del_plot("24030945", "CDKN2A",
              interval_input = 10000)
```

{{< pagebreak >}}

## PMS2

Fine setting (but not coarse) calls an exonic deletion in PMS2 in this sample which is not seen on WGS. This indicates that fine is not appropriate for larger genes.

```{r}
#| echo: FALSE

make_del_fine_plot("24030945", "PMS2",
              interval_input = 10000,
              buffer_input = 50000,
              ymin_input = -5)

```

{{< pagebreak >}}

## PTEN

WGS shows PTEN homozygous deletion on a background of monosomy 10. The coarse setting misses the focal PTEN deletion, but this is detected on fine.

```{r}
#| echo: FALSE

make_del_plot("24030945", "PTEN",
              interval_input = 200000, buffer_input = 10000,
              ymin_input = -10)

```

{{< pagebreak >}}

## BMPR1A

BMPR1A is also on chromosome 10. WGS shows a heterzygous loss (domain 3). Both coarse and fine detect this on PanSolid, with fine showing a narrower call.

```{r}
#| echo: FALSE

make_del_plot("24030945", "BMPR1A",
              interval_input = 10000000, buffer_input = 100000,
              ymin_input = -5)

```

# 24030962

This is a tetraploid sample with WGS, extracted via QIAsymphony.

```{r}
#| echo: FALSE

make_del_plot("24030962", "PTEN",
              interval_input = 10000, buffer_input = 10000,
              ymin_input = -10)

```

## BMPR1A

Both fine and coarse call a large deletion in BMPR1A. On WGS, chromosome 10 has a copy number of 2 on a background of tetraploidy.

```{r}
# echo: FALSE

make_del_plot("24030962", "BMPR1A",
                   interval_input = 2000000, buffer_input = 100000,
              ymin_input = -3)

```

## RAD51D

Fine calls a large deletion in RAD51D. This deletion is not seen on WGS.

```{r}
# echo: FALSE

make_del_fine_plot("24030962", "RAD51D",
                   interval_input = 100000, buffer_input = 100000,
              ymin_input = -5)

```

## PALB2

Both coarse and fine call a large deletion including PABL2. PABL2 has a copy number of 2 on WGS against a background of tetraploidy.

```{r}
# echo: FALSE

make_del_plot("24030962", "PALB2",
                   interval_input = 2000000, buffer_input = 100000,
              ymin_input = -3)

```

# 24025211

## CDKN2A

```{r}
#| echo: FALSE

make_del_plot("24025211", "CDKN2A",
              interval_input = 5000000, 
              buffer_input = 20000000,
              ymin_input = -3)
```

## BRCA2

On WGS, chromosome 13 has a copy number of 1. Both coarse and fine detect this deletion, but fine splits the call.

```{r}
#| echo: FALSE

make_del_plot("24025211", "BRCA2",
              interval_input = 5000000, 
              buffer_input = 5000000,
              ymin_input = -3)
```

# 24030951

WGS shows a tetraploid genome.

## PTEN

WGS shows a PTEN copy number of 0 against a background of chromosome 10 with a copy number of 2, which is against a background of a tetraploid genome.

```{r}
# echo: FALSE

make_del_plot("24030951", "PTEN",
              ymin_input = -10,
              buffer_input = 20000,
              interval_input = 20000)
```

## BMPR1A

On WGS, BMPR1A has a copy number of 2.

```{r}
# echo: FALSE

make_del_plot("24030951", "BMPR1A",
              ymin_input = -3,
              buffer_input = 20000,
              interval_input = 500000)
```

# 24035924

WGS shows CDKN2A loss on a background of 9p loss.

```{r}
# echo: FALSE

make_del_plot("24035924", "CDKN2A", interval_input = 10000)

```

## PTEN

WGS shows a loss of chromosome 10, with a PTEN copy number of 1.

```{r}
# echo: FALSE

make_del_plot("24035924", "PTEN", interval_input = 10000000,
              ymin_input = -3)

```

## CHEK2

WGS also shows a loss of chromosome 22. Both coarse and fine detect this.

```{r}
# echo: FALSE

make_del_plot("24035924", "CHEK2", interval_input = 10000000,
              ymin_input = -3)

```

# 24025266

## CDKN2A

WGS calls CDKN2A as just LOH, but CDKN2A as a biallelic loss.The coarse and fine calls both don't include all of CDKN2A. This is interesting as it indicates that not all deletions at this locus include both genes.

```{r}
# echo: FALSE

make_del_plot("24025266", "CDKN2A", interval_input = 10000,
              buffer_input = 50000,
              ymin_input = -3)

```

```{r}
#| echo: FALSE

cdkn2a_24025266 <- make_fold_change_cnv_plot(df = pansolid_del_coords |> 
                            filter(labno == "24025266"),
                          gene = "CDKN2A",
                          interval = 10000,
                          buffer = 1000000,
                          ymin = -3, ymax = 0,
                          title = "CDKN2A call zoomed in")

cdkn2a_24025266_zoom <- cdkn2a_24025266[[4]] +
  coord_cartesian(xlim = c(21978976 -100000, 22206994 -100000))

cdkn2a_24025266_exon <- make_exon_plot(21978976 -100000, 
                                       22206994 -100000,
                                       interval = 10000,
                                       chromosome = "9")

(cdkn2a_24025266_zoom / cdkn2a_24025266_exon) +
  plot_layout(heights = c(6,2))

```

## Missed het deletions

This sample has multiple chromosome losses on WGS which are not called on PanSolid, including PTEN (chr10) and CHEK2 (chr22).

# 24042543

Hexaploid genome.

## CDKN2A

```{r}
#| echo: FALSE

make_del_plot("24042543", "CDKN2A")

```

## PTEN

Fine call for an intragenic deletion of PTEN. Not seen on WGS.

```{r}
#| echo: FALSE

make_del_fine_plot("24042543", "PTEN", buffer_input = 100000,
                   interval_input = 100000,
                   ymin_input = -5)

```

## BRCA2

A large BRCA2 deletion is also called, but on WGS it has a copy number of 3. It is a relative loss comparative to the genome ploidy.

```{r}
#| echo: FALSE

make_del_plot("24042543", "BRCA2", buffer_input = 1000000,
                   interval_input = 1000000,
                   ymin_input = -5)

```

# 24048159

## CDKN2A

A CDKN2A deletion is detected on coarse but not fine. This deletion is seen on WGS.

```{r}

make_del_fine_plot("24048159", "CDKN2A")

```

```{r}

make_del_fine_plot("24048159", "STK11",
                   interval_input = 1000000,
                   buffer_input = 100000,
                   ymin_input = -5)

```

```{r}

make_del_fine_plot("24048159", "CHEK2",
                   interval_input = 1000000,
                   buffer_input = 100000, ymin_input = -5)

```

