---
title: "PanSolid Deletions Cohort Analysis"
author: "Joe Shaw (CS20980)"
date: today
date-format: "DD/MM/YYYY"
format: pdf
fontsize: 10pt
editor: visual
fig-width: 7
fig-height: 6
---

# Summary

- PanSolid can miss cases where chromosomes have a copy number of 1 (examples: 24018922)

- Non-diploid samples are tricky. PanSolid detects loss from the tumour ploidy, not absolute copy number. This means PanSolid may flag a deletion, but the gene actually has a copy number of 2 and the rest of the genome is tetraploid (example: 23033921)


```{r}
#| label: scripts-and-packages
#| include: FALSE

library(tidyverse)
library(here)

source(here("functions/extract_pansolid_cnv_coordinates.R"))

source(here("scripts/collate_gene_transcript_information.R"))

source(here("functions/cnv_plot_functions.R"))

source(here("scripts/del_val_load_processed_data.R"))

```

```{r}
#| label: functions
#| include: FALSE

pansolid_del_coords <- extract_pansolid_cnv_coordinates(del_val_pansolid_ngs_collated,
                                                        cnv_region) |> 
  mutate(fold_change = round(fold_change_adjusted, 2)) |> 
  left_join(del_val_sample_patient_info |> 
              select(labno, firstname, surname, method_name, nhsno), 
            by = "labno") |> 
  filter(!is.na(fold_change)) |> 
  rename(extraction = method_name) |> 
  mutate(extraction = ifelse(extraction == "QIAsymphony_DNA_FFPE",
                             "QIAsymphony", extraction),
         cnv_kb = round(cnv_region_length / 1000, 0))

make_del_fine_plot <- function(labno_input, gene_input, interval_input = 1000,
                               buffer_input = 5000, ymin_input = -30,
                               ymax_input = 0) {
  
  data <- pansolid_del_coords |> 
    filter(labno == labno_input & graining == "fine")
  
  fine_plot <- make_cnv_triptych_plot(make_fold_change_cnv_plot(df = data,
                                    gene = gene_input,
                                    interval = interval_input,
                                    buffer = buffer_input,
                                    ymin = ymin_input,
                                    ymax = ymax_input,
                                    title = paste0(labno_input, 
                                               " ", gene_input,
                                               " (fine)")))
  return(fine_plot)
  
}

make_del_coarse_plot <- function(labno_input, gene_input, interval_input = 1000,
                                 buffer_input = 5000, ymin_input = -30,
                               ymax_input = 0) {
  
  data <- pansolid_del_coords |> 
    filter(labno == labno_input & graining == "coarse")
  
  coarse_plot <- make_cnv_triptych_plot(make_fold_change_cnv_plot(df = data,
                                gene = gene_input,
                                interval = interval_input,
                                buffer = buffer_input,
                                ymin = ymin_input,
                                ymax = ymax_input,
                                title = paste0(labno_input, 
                                               " ", gene_input,
                                               " (coarse)")))
  
  
  return(coarse_plot)
  
}

make_del_plot <- function(labno_input, gene_input, interval_input = 1000,
                          buffer_input = 5000, ymin_input = -30, 
                          ymax_input = 0) {
  
  
  fine_plot <- make_del_fine_plot(labno_input = labno_input,
                                  gene_input = gene_input,
                                  interval_input = interval_input,
                                  buffer_input = buffer_input,
                                  ymin_input = ymin_input,
                                  ymax_input = ymax_input)
  
  coarse_plot <- make_del_coarse_plot(labno_input = labno_input,
                                  gene_input = gene_input,
                                  interval_input = interval_input,
                                  buffer_input = buffer_input,
                                  ymin_input = ymin_input,
                                  ymax_input = ymax_input)
  
  output_plot <- coarse_plot | fine_plot
    
  return(output_plot)
  
}

gene_coord_table <- read_csv(file = paste0(config::get("data_filepath"), 
                                      "validation/DOC6283_amplifications/",
                                      "gene_lists/",
                                      "gene_coordinates.csv"),
                        col_types = "ccccdd")

del_genes <- read_csv(file = paste0(config::get("data_filepath"),
                                   "validation/DOC6567_deletions/gene_lists/",
                                   "pansolid_deletion_gene_list.csv"))

```

{{< pagebreak >}}

# WGS samples

```{r}
#| label: summary-table-functions
#| include: FALSE

summarise_del_calls <- function(labno) {
  
  tbl <- pansolid_del_coords |>
    filter(labno == {{ labno }}) |> 
    select(firstname, surname, labno, extraction, graining, 
           chromosome, gene, fold_change, cnv_kb) |> 
    arrange(gene)
  
  return(tbl)
  
}

del_val_wgs_html_cnvs_with_ids <- del_val_wgs_html_cnvs |> 
  left_join(del_val_wgs_html_ids,
            by = "filepath", relationship = "many-to-one")

get_wgs_results <- function(labno) {
  
  df <- del_val_wgs_html_cnvs_with_ids |> 
    filter(labno == {{ labno }} & gene %in% del_genes$gene &
             cnv_class %in% c("LOSS", "DEL")) |> 
    select(patient_name, labno, chromosome, gene, cnv_class, cnv_copy_number)
  
  return(df)
  
}

```

## 24:H393U

### 24018922

PanSolid calls a loss of SMAD4 on both coarse and fine, and this is also detected by WGS as a loss of chromosome 18. But WGS also calls deletions on chromosomes 16 and 19, which are not called on PanSolid.

```{r}
#| label: tbl-24018922-pansolid
#| tbl-cap: "24018922 PanSolid results"
#| echo: FALSE

knitr::kable(summarise_del_calls("24018922"))

```

```{r}
#| label: tbl-24015386-wgs
#| tbl-cap: "WGS results"
#| echo: FALSE

knitr::kable(get_wgs_results("24015386"))

```

```{r}
#| echo: FALSE

make_del_plot("24018922", "SMAD4",
              interval_input = 10000000,
              buffer_input = 10000000,
              ymin_input = -3)

```

{{< pagebreak >}}

## N,23.436

### 23024556

PanSolid detects losses in CDKN2A/B and PTEN which are detected on WGS. PanSolid also detects a small PMS2 deletion - likely false positive.

```{r}
#| echo: FALSE

knitr::kable(summarise_del_calls("23024556"))

```

```{r}
#| echo: FALSE

knitr::kable(get_wgs_results("23024556"))

```

```{r}
#| echo: FALSE

make_del_plot("23024556", "CDKN2A")

```

```{r}
#| echo: FALSE

make_del_plot("23024556", "PTEN", interval_input = 10000,
              buffer_input = 10000)


```

```{r}
#| echo: FALSE

make_del_plot("23024556", "BMPR1A", interval_input = 10000000,
              buffer_input = 10000000)

```

```{r}
#| echo: FALSE

make_del_plot("23024556", "PMS2", buffer_input = 50000,
              interval_input = 20000)

```


### 24030945

```{r}
#| echo: FALSE

knitr::kable(summarise_del_calls("24030945"))

```

```{r}
#| echo: FALSE

make_del_plot("24030945", "CDKN2A",
              interval_input = 10000)
```

{{< pagebreak >}}

```{r}
#| echo: FALSE

make_del_fine_plot("24030945", "PMS2",
              interval_input = 10000,
              buffer_input = 50000,
              ymin_input = -5)

```

{{< pagebreak >}}

```{r}
#| echo: FALSE

make_del_plot("24030945", "PTEN",
              interval_input = 200000, buffer_input = 10000,
              ymin_input = -10)

```

{{< pagebreak >}}

```{r}
#| echo: FALSE

make_del_plot("24030945", "BMPR1A",
              interval_input = 10000000, buffer_input = 100000,
              ymin_input = -5)

```

### 24030946

```{r}
#| echo: FALSE

knitr::kable(summarise_del_calls("24030946"))

```

```{r}
#| echo: FALSE

make_del_plot("24030946", "CDKN2A", interval_input = 10000)

```

```{r}
#| echo: FALSE

make_del_plot("24030946", "PTEN", interval_input = 10000,
              buffer_input = 10000)

```


```{r}
#| echo: false

make_del_fine_plot("24030946", "PMS2", buffer_input = 50000, 
                   interval_input = 50000)

```


### 23023889

```{r}
#| echo: FALSE

knitr::kable(summarise_del_calls("23023889"))

```

```{r}
#| echo: FALSE

make_del_plot("23023889", "CDKN2A")

```

```{r}
#| echo: FALSE

make_del_plot("23023889", "PTEN", buffer_input = 20000, 
              interval_input = 100000)

```

## N,23.0607

Tetraploid tumour sample

### 23033921

```{r}
#| echo: FALSE

knitr::kable(summarise_del_calls("23033921"))

```

```{r}
#| echo: FALSE

knitr::kable(get_wgs_results("23033921"))

```


```{r}
#| echo: FALSE

make_del_plot("23033921", "PTEN", buffer_input = 10000,
              interval_input = 10000)


```

```{r}
#| echo: FALSE

make_del_plot("23033921", "PALB2", buffer_input = 1000000,
              interval_input = 1000000)

```


```{r}
#| echo: FALSE

make_del_plot("23033921", "BMPR1A", buffer_input = 1000000,
              interval_input = 1000000)

```

### 24030962

```{r}
#| echo: FALSE

knitr::kable(summarise_del_calls("24030962"))

```

```{r}
#| echo: FALSE

make_del_plot("24030962", "PTEN",
              interval_input = 10000, buffer_input = 10000,
              ymin_input = -10)

```


```{r}
# echo: FALSE

make_del_plot("24030962", "BMPR1A",
                   interval_input = 2000000, buffer_input = 100000,
              ymin_input = -3)

```


```{r}
# echo: FALSE

make_del_fine_plot("24030962", "RAD51D",
                   interval_input = 100000, buffer_input = 100000,
              ymin_input = -5)

```


```{r}
# echo: FALSE

make_del_plot("24030962", "PALB2",
                   interval_input = 2000000, buffer_input = 100000,
              ymin_input = -3)

```

### 23032153

```{r}
#| echo: FALSE

knitr::kable(summarise_del_calls("23032153"))

```


# 24025211

## CDKN2A

```{r}
#| echo: FALSE

make_del_plot("24025211", "CDKN2A",
              interval_input = 5000000, 
              buffer_input = 20000000,
              ymin_input = -3)
```

## BRCA2

On WGS, chromosome 13 has a copy number of 1. Both coarse and fine detect this deletion, but fine splits the call.

```{r}
#| echo: FALSE

make_del_plot("24025211", "BRCA2",
              interval_input = 5000000, 
              buffer_input = 5000000,
              ymin_input = -3)
```

# 24030951

WGS shows a tetraploid genome.

## PTEN

WGS shows a PTEN copy number of 0 against a background of chromosome 10 with a copy number of 2, which is against a background of a tetraploid genome.

```{r}
# echo: FALSE

make_del_plot("24030951", "PTEN",
              ymin_input = -10,
              buffer_input = 20000,
              interval_input = 20000)
```

## BMPR1A

On WGS, BMPR1A has a copy number of 2.

```{r}
# echo: FALSE

make_del_plot("24030951", "BMPR1A",
              ymin_input = -3,
              buffer_input = 20000,
              interval_input = 500000)
```

# 24035924

WGS shows CDKN2A loss on a background of 9p loss.

```{r}
# echo: FALSE

make_del_plot("24035924", "CDKN2A", interval_input = 10000)

```

## PTEN

WGS shows a loss of chromosome 10, with a PTEN copy number of 1.

```{r}
# echo: FALSE

make_del_plot("24035924", "PTEN", interval_input = 10000000,
              ymin_input = -3)

```

## CHEK2

WGS also shows a loss of chromosome 22. Both coarse and fine detect this.

```{r}
# echo: FALSE

make_del_plot("24035924", "CHEK2", interval_input = 10000000,
              ymin_input = -3)

```

# 24025266

## CDKN2A

WGS calls CDKN2A as just LOH, but CDKN2A as a biallelic loss.The coarse and fine calls both don't include all of CDKN2A. This is interesting as it indicates that not all deletions at this locus include both genes.

```{r}
# echo: FALSE

make_del_plot("24025266", "CDKN2A", interval_input = 10000,
              buffer_input = 50000,
              ymin_input = -3)

```

```{r}
#| echo: FALSE

cdkn2a_24025266 <- make_fold_change_cnv_plot(df = pansolid_del_coords |> 
                            filter(labno == "24025266"),
                          gene = "CDKN2A",
                          interval = 10000,
                          buffer = 1000000,
                          ymin = -3, ymax = 0,
                          title = "CDKN2A call zoomed in")

cdkn2a_24025266_zoom <- cdkn2a_24025266[[4]] +
  coord_cartesian(xlim = c(21978976 -100000, 22206994 -100000))

cdkn2a_24025266_exon <- make_exon_plot(21978976 -100000, 
                                       22206994 -100000,
                                       interval = 10000,
                                       chromosome = "9")

(cdkn2a_24025266_zoom / cdkn2a_24025266_exon) +
  plot_layout(heights = c(6,2))

```

## Missed het deletions

This sample has multiple chromosome losses on WGS which are not called on PanSolid, including PTEN (chr10) and CHEK2 (chr22).

# 24042543

Hexaploid genome.

## CDKN2A

```{r}
#| echo: FALSE

make_del_plot("24042543", "CDKN2A")

```

## PTEN

Fine call for an intragenic deletion of PTEN. Not seen on WGS.

```{r}
#| echo: FALSE

make_del_fine_plot("24042543", "PTEN", buffer_input = 100000,
                   interval_input = 100000,
                   ymin_input = -5)

```

## BRCA2

A large BRCA2 deletion is also called, but on WGS it has a copy number of 3. It is a relative loss comparative to the genome ploidy.

```{r}
#| echo: FALSE

make_del_plot("24042543", "BRCA2", buffer_input = 1000000,
                   interval_input = 1000000,
                   ymin_input = -5)

```

# 24048159

## CDKN2A

A CDKN2A deletion is detected on coarse but not fine. This deletion is seen on WGS.

```{r}

make_del_fine_plot("24048159", "CDKN2A")

```

```{r}

make_del_fine_plot("24048159", "STK11",
                   interval_input = 1000000,
                   buffer_input = 100000,
                   ymin_input = -5)

```

```{r}

make_del_fine_plot("24048159", "CHEK2",
                   interval_input = 1000000,
                   buffer_input = 100000, ymin_input = -5)

```


# Repeatability Samples



```{r}

inter_run_replicates <- c(
  # 24017319
  "24017319_WS141734", "24017319a_WS144291", "24017319_WS144364",
  # 24027774
  "24027774_WS142076", "24027774a_WS144291", "24027774_WS144364",
  # 24020998
  "24020998_WS141045", "24020998a_WS144291", "24020998_WS144364")

intra_run_replicates <- c(
  # 24017319
  "24017319a_WS144291", "24017319b_WS144291", "24017319c_WS144291",
  # 24027774
  "24027774a_WS144291", "24027774b_WS144291", "24027774c_WS144291",
  # 24020998
  "24020998a_WS144291", "24020998b_WS144291", "24020998c_WS144291"
)

all_replicates <- c(inter_run_replicates, intra_run_replicates)

repeat_data <- del_val_pansolid_ngs_collated |> 
  mutate(labno_suffix_worksheet = paste0(labno, suffix, "_", worksheet)) |> 
  filter(labno_suffix_worksheet %in% all_replicates) |> 
  select(graining, labno, worksheet, labno_suffix_worksheet, gene, fold_change_adjusted) 

repeat_data |> 
  filter(labno == "24020998") |> 
  ggplot(aes(x = labno_suffix_worksheet, y = fold_change_adjusted)) +
  geom_col(aes(fill = gene), position = "dodge") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(x = "", y = "Fold Change") +
  facet_wrap(~graining)

```

# ddPCR Samples

