---
title: "INC10886: MSH2 variant case"
author: "Joe Shaw (CS20980)"
format:
  html:
    code-fold: true
    code-summary: "Show the code"
    embed-resources: true
editor: visual
---

```{r}
#| label: packages
#| include: FALSE

library(tidyverse)
library(readxl)
library(here)

source(here("functions/pansolid_cnv_excel_functions.R"))
source(here("functions/chromosome_arm_functions.R"))

```

# Introduction

Sample 25038352 had a potentially pathogenic MSH2 variant, a `r str_length("CTTCGTTCTGACTTCTCCAAGTTTCAGGAAATGAT")`bp deletion in exon 9, detected on WS155167 (G02) which then was not detected on WS155702 (F02) or WS156308 (G02).

Variant: MSH2 c.1327_1361delCTTCGTTCTGACTTCTCCAAGTTTCAGGAAATGAT p.(Leu443fs) 25% 80 out of 323 reads on WS155167

Panel: v2bM1_tLYNCH_PS

Chromosome region: 47445598..47445632

Target: chr2 47445543..47445662, target number 637, which is 120bp long.

# Methods

Target level data for every sample on each worksheet (WS155167, WS155702 and WS156308) was collated.
This step was computationally quite intensive so I have exported the collated data into the incident folder for future reference.

```{r}
#| label: collate-data
#| eval: FALSE

# Worksheet where MSH2 c.1327_1361del variant was detected
WS155167_files <- list.files(path = paste0(config::get("worksheet_folderpath"),
                                           "WS155167"), full.names = TRUE,
           recursive = TRUE, pattern = "Results_SNVs_Indels.*.xlsx")

# Variant not detected
WS156308_files <- list.files(path = paste0(config::get("worksheet_folderpath"),
                                           "WS156308"), full.names = TRUE,
           recursive = TRUE, pattern = "Results_SNVs_Indels.*.xlsx")

# Variant not detected
WS155702_files <- list.files(path = paste0(config::get("worksheet_folderpath"),
                                           "WS155702"), full.names = TRUE,
                             recursive = TRUE, pattern = "Results_SNVs_Indels.*.xlsx")

# Worksheets with high signal-adjusted noise - for later analysis
WS153211_files <- list.files(path = paste0(config::get("worksheet_folderpath"),
                                           "WS153211"), full.names = TRUE,
                             recursive = TRUE, pattern = "Results_SNVs_Indels.*.xlsx")

WS154828_files <- list.files(path = paste0(config::get("worksheet_folderpath"),
                                           "WS154828"), full.names = TRUE,
                             recursive = TRUE, pattern = "Results_SNVs_Indels.*.xlsx")

WS157371_files <- list.files(path = paste0(config::get("worksheet_folderpath"),
                                           # Use the original run data only
                                           "WS157371/old"), full.names = TRUE,
                             recursive = TRUE, pattern = "Results_SNVs_Indels.*.xlsx")

ws_files <- c(WS155167_files, 
              WS156308_files, 
              WS153211_files, 
              WS154828_files, 
              WS157371_files)

# Remove any temporary files created when the Excels are opened
ws_files <-  ws_files[grepl(
  pattern = "^(?!.*~\\$).*\\.xlsx$",
  x = ws_files,
  perl = TRUE)]

# Each worksheet should only have ~48 files
# Make sure original and rerun data are not accidentally collated together
stopifnot(length(ws_files) / 5 < 50)

ws_data <- ws_files |> 
  map(\(ws_files) read_targets_merged(filepath = ws_files, 
                                      sheetname = "CNV Targets Merged")) |> 
  list_rbind()

# Have to collate WS155702 separately as the sheet name is different
WS155702_data <- WS155702_files |> 
  map(\(WS155702_files) read_targets_merged(filepath = WS155702_files, 
                                      sheetname = "CNV Targets Merged-1")) |> 
  list_rbind()

all_data <- rbind(ws_data, WS155702_data)

all_data_mod <- extract_pansolid_cnv_coordinates(all_data, 
                                                 cnv_coord_col = region) |> 
  mutate(chromosome = str_replace(chromosome, "\\.0", ""),
         chromosome = str_c("chr", chromosome),
         target_coord_string = str_c(chromosome, "_",
                               start, "-", end))

write_csv(all_data_mod, paste0(config::get("data_folderpath"),
                           "live_service/INC10886_MSH2_variant/",
                           "collated_target_data.csv"))

```

```{r}
#| label: load-data

all_data <- read_csv(file = paste0(config::get("data_folderpath"),
                           "live_service/INC10886_MSH2_variant/",
                           "collated_target_data.csv"),
                     show_col_types = FALSE)

```

# Results

## WS155167 quality

```{r}
#| label: identify-low-regions
#| message: FALSE

grouped_by_region <- all_data |> 
  group_by(worksheet, chromosome, target_coord_string) |> 
  summarise(
    max_fc = max(fold_change_adjusted, na.rm = TRUE),
    median_fc = median(fold_change_adjusted),
    min_fc = min(fold_change_adjusted, na.rm = TRUE),
    median_cov = median(case_coverage)) |> 
  mutate(cov_category = case_when(
    median_cov >= 138 ~"ok",
    median_cov < 138 ~"low"
  ))

low_regions_ws155167 <- grouped_by_region |> 
  filter(worksheet == "WS155167" & median_cov < 138)

```

WS155167 has increased signal-adjusted noise across every sample on the worksheet.
There are `r nrow(low_regions_ws155167)` targets on WS155167 which have a median coverage of less than 138X.
The root cause of why certain targets would have consistently lower coverage for all samples is not clear.
It was hypothesised that lower coverage for the target overlapping the *MSH2* variant (target number 637) on WS155167 could have somehow caused the variant to be detected.
However, the median coverage of target 637 on WS155167was not low in comparison to other targets on WS155167 or in comparison to the same target on WS155702 and WS156308.

```{r}
#| label: all-targets-plot
#| echo: FALSE

all_targets_plot <- grouped_by_region |> 
                             filter(worksheet %in% c("WS155167", 
                                                     "WS155702", 
                                                     "WS156308")) |> 
  ggplot(aes(x = target_coord_string, y = median_cov)) +
  geom_point(shape = 21, aes(fill = cov_category))+
  scale_fill_manual(values = c( "red", "white")) +
  geom_hline(yintercept = 138, linetype = "dashed") +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.position = "none") +
  facet_wrap(~worksheet, ncol = 1) +
  labs(y = "Median coverage across samples", x = "Target",
       title = paste0("There are ",
                      nrow(low_regions_ws155167),
                      " targets with median coverage below 138X on WS155167"),
       subtitle = "Data shown for all samples on each worksheet. Target 637 in blue",
       caption = "Each dot represents a median value across all samples") +
  coord_cartesian(ylim = c(0, 1000)) +
  scale_y_continuous(breaks = c(0, 138, 500, 750, 1000)) +
  geom_point(data = grouped_by_region |> 
               filter(worksheet %in% c("WS155167", 
                                       "WS155702", 
                                       "WS156308")) |> 
               # Target 637
               filter(target_coord_string == "chr2_47445543-47445662"), 
             shape = 21, size = 4, 
             fill = "blue"
             )

all_targets_plot

```

It was then hypothesised that, whilst the median coverage of target 637 was not low, the coverage in sample 25038352 on worksheet WS155167 was low.
However, whilst the general coverage pattern for 25038352 seems to be in the lower range of samples on WS155167, target 637 is not particularly impacted.

```{r}
#| label: target-637
#| echo: FALSE

target_637_plot <- all_data |> 
  filter(target_number > 620 &
           target_number < 650) |> 
  filter(worksheet %in% c("WS155167", 
                          "WS155702", 
                          "WS156308")) |> 
  mutate(target_number = as.character(target_number)) |> 
  ggplot(aes(x = target_number, y = case_coverage)) +
  geom_boxplot() +
  theme_bw() +
  geom_point(data = all_data |> 
               filter(labno == "25038352" &
                         target_number > 620 &
           target_number < 650) |> 
             mutate(target_number = as.character(target_number)), 
             shape = 21, size = 4, 
             fill = "red",
             ) +
  labs(x = "", y = "Coverage",
       title = "Coverage of targets 620-650",
       subtitle = "Sample 25038352 in red, target 637 highlighted in blue") +
  facet_wrap(~worksheet, ncol = 1) +
   annotate('rect', xmin = "636", xmax = "638", ymin = 0, 
             ymax = 3000, fill= "blue", alpha = 0.2) +
  coord_cartesian(ylim = c(0, 2500))

target_637_plot

```

## GC content: primers

The cause of certain targets having poor coverage across all samples on WS155167 was then investigated.
It was hypothesised that primer GC content could impact coverage, potentially due to a sub-optimal thermocycling step.
However, there does not seem to be a relationship between primer GC content and median coverage on WS155167.

```{r}
#| label: gc-content-primers
#| echo: FALSE

primer_target_df <- read_csv(paste0(config::get("data_folderpath"),
                                    "validation/DOC6791_chromosome_arms/",
                                    "primers_and_targets/",
                                    "primer_target_df.csv"),
                             show_col_types = FALSE)

ps_primers_grch38 <- read_csv(paste0(config::get("data_folderpath"),
                                    "validation/DOC6791_chromosome_arms/",
                                    "primers_and_targets/",
                                    "ps_primers_grch38.csv"),
                              show_col_types = FALSE)

primer_target_df_gc <- primer_target_df |> 
  left_join(ps_primers_grch38, by = "primer_coord_str")
  
primer_target_median_gc <- primer_target_df_gc |> 
  group_by(target_coord_str) |> 
  summarise(median_primer_gc = median(primer_gc_percent),
            median_length = median(primer_length))

all_data_with_primer_gc <- all_data |> 
  left_join(primer_target_median_gc,
            join_by(target_coord_string == target_coord_str))

grouped_by_region_with_primer_gc <- grouped_by_region |> 
  left_join(primer_target_median_gc,
            join_by(target_coord_string == target_coord_str))

grouped_by_region_with_primer_gc |> 
  filter(worksheet == "WS155167") |>
  filter(!is.na(median_primer_gc)) |> 
  mutate(cov_category = case_when(
    median_cov < 138 ~"Below 138X",
    median_cov >= 138 ~"Above 138X"
  )) |> 
  ggplot(aes(x = median_primer_gc, y = )) +
  geom_histogram(aes(fill = cov_category),
                 colour = "black",
                 binwidth = 1) +
  labs(x = "Median primer GC percent",
       y = "Number of targets",
       fill = "Median coverage",
       title = "Primer GC content and coverage on WS155167") +
  theme_bw()

```

## GC content: target regions of interest

The same relationship was hypothesised for the target regions, however there is no evidence that target GC percent influences the median coverage of a target.

```{r}
#| label: gc-content-targets
#| echo: FALSE

ps_grch38_fasta <- read_delim(file = paste0(config::get("data_folderpath"),
                                            "validation/DOC6791_chromosome_arms/",
                                            "bed_files/",
                                            "QIAseq_DNA_panel.CDHS-48608Z-11752.roi_GRCh38_fasta.bed"),
                              delim = "\t",
                              col_names = c("chrom", "start", "stop",
                                            "name", "sequence"),
                              show_col_types = FALSE)

ps_grch38_fasta_mod <- ps_grch38_fasta |> 
  mutate(target_length = str_length(sequence),
         gc_count = str_count(string = sequence,
                                pattern = "G|C"),
         gc_content = (gc_count / target_length) * 100,
         start_plus_one = start +1,
         target_region = str_c(chrom, "_",
                               start_plus_one, "-", stop))

grouped_by_region_gc <- grouped_by_region |> 
  left_join(ps_grch38_fasta_mod,
            join_by("target_coord_string" == "target_region"),
            relationship = "many-to-one")

grouped_by_region_gc |> 
  filter(worksheet == "WS155167" & !is.na(gc_content)) |> 
  filter(target_length > 11) |> 
  mutate(cov_category = case_when(
    median_cov < 138 ~"Below 138X",
    median_cov >= 138 ~"Above 138X"
  )) |> 
  ggplot(aes(x = gc_content, y = )) +
  geom_histogram(aes(fill = cov_category),
                 colour = "black",
                 binwidth = 1) +
  labs(x = "Median target GC percent",
       y = "Number of targets",
       fill = "Median coverage",
       title = "Target GC content and coverage on WS155167") +
  theme_bw()

```

## Targets on other worksheets with high signal-adjusted noise

We then investigated whether the targets impacted by low coverage in every sample on WS155167 were also impacted on 3 other worksheets with high signal-adjusted noise: WS157371, WS153211 and WS154828.

```{r}
#| label: tbl-worksheet-notes
#| echo: FALSE

worksheet_notes <- tribble(
  ~"Worksheet", ~"Noise", ~"Sample 25038352 result", ~"Was the sequencing repeated?",
  "WS155167", "High", "MSH2 c.1327_1361del variant detected ", "No",
  "WS156308", "Low", "MSH2 c.1327_1361del variant NOT detected", "No",
  "WS155702", "Low", "MSH2 c.1327_1361del variant NOT detected", "No",
  "WS157371", "High", "Not tested", "Yes",
  "WS154828", "High", "Not tested", "No",
  "WS153211", "High", "Not tested", "Yes"
)

knitr::kable(worksheet_notes)

```

There were only 37 targets which had consistently low coverage across the 4 worksheets.
It was noted that there was a strong overlap of targets between WS155167 and WS157371, and a strong overlap between WS154828 and WS153211.

```{r}
#| label: target-venn-diagram
#| echo: FALSE

WS155167_low_targets <- grouped_by_region |> 
  filter(median_cov < 138 & worksheet== "WS155167") 

WS157371_low_targets <- grouped_by_region |> 
  filter(median_cov < 138 & worksheet== "WS157371") 

WS153211_low_targets <- grouped_by_region |> 
  filter(median_cov < 138 & worksheet== "WS153211") 

WS154828_low_targets <- grouped_by_region |> 
  filter(median_cov < 138 & worksheet== "WS154828")

low_target_comparison <- list(
  "WS155167" = WS155167_low_targets$target_coord_string,
  "WS157371" = WS157371_low_targets$target_coord_string,
  "WS153211" = WS153211_low_targets$target_coord_string,
  "WS154828" = WS154828_low_targets$target_coord_string
)

ggvenn::ggvenn(low_target_comparison,
               set_name_size = 3,
               show_percentage = FALSE) +
  labs(title = "Targets with median coverage below 138X",
       subtitle = "On worksheets with high signal-adjusted noise")

```

# Conclusion

The cause of the reduced coverage for `r nrow(low_regions_ws155167)` targets on WS155167 has not been found.
The cause of the MSH2 c.1327_1361delCTTCGTTCTGACTTCTCCAAGTTTCAGGAAATGAT being detected for sample 25038352 on WS155167 but not on WS155702 and WS156308 has not been found.
