---
title: "INC10378"
author: "Joe Shaw (CS20980)"
format:
  html:
    code-fold: true
    code-summary: "Show the code"
    embed-resources: true
editor: visual
---

```{r}
#| label: packages-and-functions
#| include: FALSE

library(tidyverse)
library(readxl)
library(janitor)

source(here::here("functions/pansolid_cnv_excel_functions.R"))
source(here::here("functions/chromosome_arm_functions.R"))
source(here::here("functions/contamination_functions_INC10994.R"))

```

# Introduction

Index contamination was seen on WS151383 for well C3 (25015163 - Endometrial cancer) and well C4 (25014769 - Lung cancer).

On the same worksheet, sample 25015130 was also noted to have a SNP profile consistent with contamination.

# Methods

SNP data from PanSolid WS151383 was reviewed.

```{r}
#| label: load-data

WS151383_data <- collate_snp_data("WS151383")

```

MSI testing was performed on WS155852.

# Results

## 25014769

### Was 25014769 contaminated with 25015163?

Sample 25014769 (lung) was suspected to be contaminated at the indexing stage with sample 25015163 (endometrial).
The plot below shows that this is likely the case.

```{r}
#| label: lung-plot
#| echo: FALSE
#| fig-width: 6
#| fig-height: 6
#| warning: FALSE

make_contamination_snp_plot(WS151383_data,
                            "25014769_WS151383",
                           WS151383_data,
                           "25015163_WS151383",
                           het_vaf_upper_threshold = 80,
                           het_vaf_lower_threshold = 20)

```

### What were the SNP results on Genexus testing?

25014769 was also tested using the Genexus.
We can compare the SNP frequencies for SNPs detected on Genexus against those detected on PanSolid.
In order to do this we have to convert the Genexus SNP data from GRCh37 from GRCh38 using the [Ensembl Assembly Converter](https://grch37.ensembl.org/Homo_sapiens/Tools/AssemblyConverter).

```{r}
#| label: genexus-pansolid-snps

snp_25014769 <- read_excel(path = paste0(
  "S:/central shared/Genetics/Repository/",
  "WorksheetAnalysedData/WS151383/v2aM4_LUNG_PS/",
  "Unannotated_Files/",
  "Results_SNVs_Indels-WS151383_25014769_S16_R1_001.xlsx"),
  sheet = "Artefacts_removed_GIAB_filt...",
  range = cell_cols("A:L"),
  col_types = c(
               "numeric", "text", "text", "text", "text", 
               "text", "numeric", "text", "text",
               "numeric", "numeric", "numeric")) |>  
  janitor::clean_names() |> 
  filter(type == "SNV") |> 
  mutate(region = as.numeric(region))

# Export from Genexus software
genexus_data <- read_excel(path = paste0(config::get("data_folderpath"),
                           "live_service/INC10378_indexes/",
                           "25014769_r25-0e61_variants.xlsx")) |> 
  janitor::clean_names()

# Reformat as a bed file
genexus_bed <- genexus_data |> 
  mutate(chrom = str_extract(string = locus,
                             pattern = "chr(\\d{1,2}|X)"),
         start = str_extract(string = locus,
                             pattern = ":(\\d{1,10})",
                             group = 1),
         end = start) |> 
  select(chrom, start, end)

write_csv(genexus_bed, "genexus_bed.csv")

# Output from Assembly Converter
genexus_grch38 <- read_tsv(file = paste0(config::get("data_folderpath"),
                           "live_service/INC10378_indexes/",
                           "Em3KLKrw7mCzZvKG.txt"),
         col_names = c("chrom", "start", "end"),
         show_col_types = FALSE) |> 
  mutate(grch38_locus = paste0(chrom, ":", start))

genexus_mod <- genexus_data |> 
  cbind(genexus_grch38) |> 
  rename(grch37_locus = locus,
         genexus_frequency = allele_frequency_percent) |> 
  select(grch37_locus, grch38_locus, alt, genexus_frequency)

genexus_ps_join <- snp_25014769 |> 
  rename(ps_frequency = frequency) |> 
  mutate(grch38_locus = paste0("chr", chromosome,
                               ":", region)) |> 
  select(grch38_locus, allele, ps_frequency) |> 
  inner_join(genexus_mod, join_by("grch38_locus" == "grch38_locus",
                                  "allele" == "alt"),
             keep = TRUE) |> 
  select(grch37_locus, grch38_locus.y, grch38_locus.x,
         allele, alt, genexus_frequency, ps_frequency)

```

There are only `r nrow(genexus_ps_join)` SNPs for comparison between the Genexus and PanSolid.
The frequencies of each SNP are broadly similar on each method, however due to the low numbers of SNPs this is a limited comparison, and it is unlikely that the Genexus data is also contaminated with 25015163.

These samples were on different QIAsymphony extraction batches (25015163, QS90979; 25014769 QS90925) on different days, and were sent from different pathology labs (25015163 Manchester Royal Infirmary; 25014769 Wythenshawe Hospital).
Therefore the likelihood of contamination in the stock DNA is very low.
MSI testing on WS155852 showed no evidence of contamination for either 25014769 or 25015163.

```{r}
#| label: genexus-pansolid-plot
#| echo: FALSE

genexus_ps_plot <- ggplot(genexus_ps_join, aes(x = ps_frequency, y = genexus_frequency)) +
  geom_point(shape = 21, size = 2) +
  geom_abline(linetype = "dashed") +
  labs(x = "PanSolid SNP frequency (%)", y = "Genexus SNP frequency (%)") +
  scale_x_continuous(limits = c(0, 100),
                     breaks = seq(0, 100, by = 10)) +
  scale_y_continuous(limits = c(0, 100),
                     breaks = seq(0, 100, by = 10)) +
  theme_bw()

genexus_ps_plot

```

## 25015130

### Where did the contamination for 25015130 come from?

MSI testing on WS155852 showed that the stock DNA for 25015130 is contaminated.
The source of contamination for 25015130 was investigated using the SNP-matching method described in INC10994.
All samples on worksheets WS151383 and WS151504 were checked.
No sample had greater than 90% matching homozygous SNPs with 25015130, which is used as a threshold for likely contamination. 

```{r}
#| label: contaminant-search-25015130
#| echo: FALSE

WS151504_data <- collate_snp_data("WS151504")

all_data_25015130 <- rbind(WS151383_data,
                           WS151504_data)

results_25015130 <- search_for_contaminant(all_data_25015130,
                                           "25015130_WS151383")

```

Sample 25015130 was extracted on QIAsymphony batch 90956. 
Not all samples on this extraction batch were tested with PanSolid NGS, so excluding that the contamination arose at DNA extraction is not possible.

```{r}
#| label: qs-batch-90956
#| echo: FALSE

knitr::kable(ps_samples_per_qs_batch("90956"))

```

Therefore the source of the contamination is unknown, and is unlikely to be from the PanSolid library preparation.
For context, a source of contamination could not be found for 24/34 contaminated samples investigated for INC10994.
This is likely due to the fact that contamination may occur during the preparation stage at a referring pathology lab, and the contaminating sample is not also sent to our lab for NGS testing.

### Should scientists have spotted contamination for 25015130?

WS151383 was run on 18th March. INC10288 was raised on 27th February.

DOC6261 revision 2 was the CNV SOP at the time: “The whole-genome HTML file (Figure 21) is an analysis aid provided to assist with interpretation of complex cases – it does not need to be opened for every sample.”

The first email sent by me to summarise the ongoing contamination investigation was on 4th April.

My notes from contamination meeting on 18th March state: “Adding a DNA contamination check can be introduced at the same time as the new CNV format, to make training simpler”

Therefore scientists were not expected to be checking for contamination at this stage.

# Conclusion

The PanSolid results for sample 25014769 are contaminated with sample 25015163.
This contamination occurred at the indexing stage of library preparation on PanSolid WS151383. 

The source of contamination for sample 25015130 is unknown. Scientists were not informed to check PanSolid HTML outputs for contamination at the time 25015130 was tested, although this has since been introduced.
