---
title: "Validation of Detecting Somatic Gene Amplifications with the PanSolid CLC Pipeline"
format: 
  docx:
    reference-doc: DOC6283_template.docx
toc: true
toc-title: "Table of Contents"
---

# Authors\

Author: Joe Shaw

Project Lead(s): Joe Shaw, Elizabeth Sollars, Eleanor Baker, Helene Schlecht, George Burghel

Date:	`r Sys.Date()`\

Author contact: joseph.shaw2@mft.nhs.uk; joseph.shaw3@nhs.net\

# Acronyms\

ALK: 			Anaplastic lymphoma kinase

CTNNB1:		Catenin beta 1

CLC:			Acronym unknown. CLC Bio was acquired by Qiagen in 2013.	

CNV:			Copy Number Variant

ddPCR:    droplet digital Polymerase Chain Reaction

EGFR:			Epidermal Growth Factor Receptor

ERBB2:			ERythroBlastic oncogene B 2

FFPE:			Formalin-Fixed Paraffin Embedded

FISH:			Fluorescent In-Situ Hybridisation

GLH:			Genomic Laboratory Hub

NGS:			Next Generation Sequencing

NICE:			National Institute for Health and Care Excellence

UMI: 			Unique Molecular Identifier

WGS:			Whole Genome Sequencing\

# Purpose\

The purpose of this document is to describe the validation of detection of somatic gene amplifications on the PanSolid v2 QIAseq panel using the PanSolid CLC pipeline at the North West Genomic Laboratory Hub (GLH).\

# Introduction\

At the North West GLH, somatic variant testing is performed using the QIAseq methodology (Qiagen) (1) and Illumina next generation sequencing (NGS) using a large panel of QIAseq primer targets named “PanSolid”. The current PanSolid enrichment, referred to as PanSolidv2 (CDHS-48608Z-11752) contains 11,752 unique QIAseq targets. Sequencing data from the PanSolid enrichment is analysed with a pipeline built in the Qiagen CLC Genomic Workbench. Detection of single nucleotide variants and indels on PanSolid has previously been validated (DOC6024 Validation of QIASeq PanSolid Enrichment). Currently, copy number variants (CNVs) are analysed using a separate QIAseq “core” panel with a bespoke pipeline written in Python, with copy number amplifications reported for 4 genes: EGFR, ERBB2, AR and MET. The long-term aim is to move all CNV detection to the PanSolid panel and the CLC pipeline.\

The detection of ERBB2 gene amplifications was validated on the PanSolid enrichment in April 2024 (DOC6260: Validation of Detecting Somatic ERBB2 Amplifications with the PanSolid CLC Pipeline). ERBB2 amplifications were prioritised for validation due to service needs, and were validated by comparison with results from droplet digital PCR testing. The aim of this validation document is to validate additional amplification genes on the PanSolid workflow; these genes are specified by the National Genomic Test Directory.\

```{r}
#| label: packages-and-data
#| warning: FALSE
#| include: FALSE

library(tidyverse)
library(knitr)
library(here)
library(patchwork)
library(janitor)

source(here("scripts/set_shared_drive_filepath.R"))

source(here("scripts/load_processed_validation_data.R"))

source(here("functions/gene_table_functions.R"))

source(here("functions/pansolid_excel_functions.R"))

source(here("functions/primer_functions.R"))

```

##	Somatic Gene Amplifications\

Gene amplification is a common mechanism in the development of cancer (2).  Additional gene copies can be present on extra-chromosomal DNA rather than their initial chromosomal loci (3), for example EGFR amplifications can be present on extrachromosomal DNA fragments referred to as “double minutes” (4). “Double minute” refers to a very small (“minute”) bodies of chromatin material with two arms (“double) which were originally visualised by chromosome staining (5).\

The number of additional gene copies can vary between genes and tumour types, ranging as high as hundreds of gene copies. There is no consistent accepted threshold of copy number to classify a gene as “amplified”, although many clinical trials use a threshold of 10 copies as an entry requirement.\

Gene amplifications can have clinical relevance for choice of treatment and predicting resistance to treatment, particularly amplifications of genes encoding tyrosine kinase membrane receptors. It is predicted that increasing the gene copy number increases the levels of translated protein, which maintains aberrant cell signalling even in the presence of tyrosine kinase inhibitors (TKIs) (6). Historically, gene amplifications were detected with fluorescence in-situ hybridisation (FISH) and immunohistochemistry (IHC), but next generation sequencing (NGS) methods are being increasingly used. The PanSolid enrichment includes 212 genes relevant to cancer. Of these, there are 9 genes which were selected as having amplifications with clinical relevance (@tbl-amplification-coordinates).\

```{r}
#| label: tbl-amplification-coordinates
#| tbl-cap: "Gene Coordinates"
#| warning: FALSE
#| echo: FALSE

primer_df <- read_csv(paste0(data_folder, 
                             "primers/QIAseq.CDHS-48608Z-11752.primers3 (Converted).csv")) |> 
  clean_names()

target_df <- read_csv(paste0(data_folder,
                             "bed_files/PanSolidv2_GRCh38_noalt_BED.csv")) |> 
  clean_names()

amp_genes <- load_pansolid_gene_table("Amplifications")

gene_tbl <- read_csv(file = paste0(data_folder, "gene_lists/gene_coordinates.csv"),
                       col_types = "ccccdd")

primer_df_mod <- extract_cnv_coordinates(df = primer_df, cnv_coord_col = region) |> 
  mutate(region_length = abs(start-end)) |> 
  rename(primer_start = start,
         primer_end = end)

gene_tbl_mod <- gene_tbl |> 
  mutate(gene_size_kb = round(abs(gene_start - gene_end) / 1000, 1)) |> 
  rowwise() |> 
  mutate(primers_in_gene = count_primers_in_region(chrom = chromosome,
                                                   coord1 = gene_start,
                                                   coord2 = gene_end,
                                                   df = primer_df_mod)) |> 
  filter(gene %in% amp_genes$gene) |> 
  arrange(gene)

gene_target_counts <- target_df |> 
  filter(name %in% amp_genes$gene) |> 
  count(name)

gene_coordinate_tbl <- gene_tbl_mod |> 
  mutate(grch38_coordinates = str_c("chr", chromosome, ":", gene_start, "-", gene_end)) |> 
  select(gene, grch38_coordinates, gene_size_kb, transcript_refseq,
         primers_in_gene) |> 
  left_join(gene_target_counts, 
            join_by("gene" == "name")) |> 
  rename(`Targets in gene` = n,
         `Gene` = gene,
         `Coordinates (GRCh38)` = grch38_coordinates,
         `Size (kb)` = gene_size_kb,
         `Transcript (RefSeq)` = transcript_refseq,
         `Primers in gene` = primers_in_gene)

kable(gene_coordinate_tbl)

```
\
```{r}
#| label: amplification-clinical-relevance
#| echo: FALSE


```

## Requirements\

The requirements for this new method are as follows:\

- **Clinical utility:** the test should provide clear results for the 10 listed amplification markers.

- **Clinical validity:** the test should have a defined sensitivity and specificity for these markers, calculated by comparison with an orthogonal test.

- **Dosage accuracy:** it should be shown how accurate the test is at detecting amplifications of different copy number states, expressed as a correlation between the test and copy numbers from an orthogonal test.

- **Quality control:** the test should have clear quality control thresholds for deciding when the test results should be reported and when the test has failed.

- **Variability:** the test should have defined intra-run variation (repeatability) and inter-run variation (repeatability) for these amplifications.

- **Visualisation:** the results should be presented in a clear and visual manner.

- **Analysis:** there should be a clear procedure for analysis of copy number variants by laboratory staff.\

# Methodology\

## Validation Structure\

## Samples\

Samples were received at the North West GLH between December 2019 and July 2024. Details of the normal control cohort used for the PanSolid CLC pipeline are given in the ERBB2 validation document (DOC6260). In total, 94 different samples were tested on the PanSolid version 2 enrichment during this validation.\

```{r}
#| label: tbl-tissue-types
#| tbl-cap: "Sample tissue types"
#| echo: FALSE

tissue_type_summary <- validation_sample_patient_info |> 
    filter(labno %in% validation_stdev_results_collated$labno) |> 
    mutate("Tissue type" = case_when(
      tissue_type == "Shavings" ~"FFPE shavings",
      tissue_type %in% c("Slides", "Positively-charged slides") ~"FFPE slides",
      TRUE ~tissue_type
    )) |> 
    count(`Tissue type`) |> 
    rename("Samples" = n) |> 
    arrange(desc(Samples)) |> 
    adorn_totals()

kable(tissue_type_summary)

```

```{r}
#| label: cancer-types-table
#| tbl-cap: "Sample cancer types"
#| echo: FALSE

cancer_type_summary <- validation_sample_patient_info |> 
  filter(labno %in% validation_stdev_results_collated$labno) |> 
  #filter(cancer_type_new != "Reference material") |> 
  count(cancer_type_new) |> 
  rename("Samples" = n,
         "Cancer" = cancer_type_new) |> 
  arrange(desc(Samples)) |> 
  adorn_totals()

kable(cancer_type_summary)

```

```{r}
#| label: extraction-types-table
#| tbl-cap: "DNA extraction methods"
#| echo: FALSE

extraction_type_summary <- validation_sample_patient_info |> 
  filter(labno %in% validation_stdev_results_collated$labno) |> 
  count(method_name) |> 
  rename("DNA extraction method" = method_name,
         "Samples" = n) |> 
  arrange(desc(Samples)) |> 
  adorn_totals()

kable(extraction_type_summary)

```

```{r}
#| label: ncc-table
#| tbl-cap: "Sample neoplastic cell contents"
#| echo: FALSE

ncc_summary <- validation_sample_patient_info |> 
    filter(labno %in% validation_stdev_results_collated$labno) |> 
    filter(neoplastic_cell_content != "Reference material") |> 
    mutate(ncc_class = case_when(
      neoplastic_cell_content %in% c(">10", "10-20%") ~">10%",
      neoplastic_cell_content %in% c("20-50", "20-30%",
                                     ">20%", ">20", 
                                     "20") ~">20%",
      neoplastic_cell_content %in% c(">30%", ">30") ~">30%",
      neoplastic_cell_content %in% c("50-75") ~">50%",
      neoplastic_cell_content %in% c( ">75", "80") ~">75%",
      TRUE ~neoplastic_cell_content
    ),
    ncc_class = factor(ncc_class, levels = c(">75%",
                                             ">50%",
                                             ">30%",
                                             ">20%",
                                             ">10%",
                                             "No NCC provided"))) |> 
    count(ncc_class) |> 
    rename("Neoplastic cell content" = ncc_class,
           "Samples" = n) |> 
    adorn_totals()

kable(ncc_summary)

```

## Reference Materials\

### SeraCare reference materials\

Reference materials were ordered from the company SeraCare to act as positive controls for different gene amplifications. Each control consists of DNA extracted from an FFPE sample of a stable cell line, spiked with plasmid DNA for specific genes at different quantities. The amplified genes included in the SeraCare reference materials are shown in Table 7.\

```{r}
#| label: tbl-seracare-reference-material-genes
#| tbl-cap: "SeraCare reference material genes"
#| echo: FALSE
#| warning: FALSE

seracare_gene_overlap <- read_csv(file = paste0(data_folder, 
                                                 "seracare_reference_materials/",
                                                 "seracare_gene_copies.csv")) |> 
  filter(reference_material_short == "WT") |> 
  select(gene) |> 
  mutate(`Targetted within this validation` = ifelse(gene %in% amp_genes$gene, "Yes", "No")) |> 
  rename(`Gene` = gene)
  
kable(seracare_gene_overlap)

```

The details of the SeraCare controls are shown in @tbl-seracare-control-details. Note that there were two tubes of both the Seraseq® TNA (DNA) WT Mix and the Seraseq® Solid Tumor CNV Mix, +12 copies mix.\

```{r}
#| label: tbl-seracare-reference-material-details
#| tbl-cap: "SeraCare reference material details"
#| echo: FALSE
#| warning: FALSE

seracare_details <- read_csv(file = paste0(data_folder,
                                           "seracare_reference_materials/",
                                           "seracare_sample_details.csv")) 

kable(seracare_details)

```

### Limit of detection DNA mixtures\

A limit of detection experiment was performed using the Seraseq® Solid Tumor CNV Mix, +12 copies reference material, which represented DNA from tumour cells, and the Seraseq® TNA (DNA) WT Mix, which represented DNA from normal cells. The two reference materials were mixed together to mimic samples with neoplastic cell contents of 30%, 20%, 10% and 0%. The final volume of each DNA mixture was 20 µl and the total DNA concentration was 6 ng/µl, which is the standard input for the PanSolid NGS workflow. These DNA mixtures were made on worksheet WS144265 and run on the PanSolid version 2 NGS workflow on worksheet WS144291. This limit of detection experiment is different to the limit of detection experiment performed on the PanSolid version 1 workflow for the validation of ERBB2 amplifications (DOC6260).\

## Next generation sequencing\

DNA was extracted using the QIASymphony extraction method (DOC6251: QIAsymphony DNA FFPE Extraction Procedure) and tested on version 2 (CDHS-48608Z-11752) of the PanSolid QIAseq enrichment (DOC5925: QIAseq PanSolid DNA Panel NGS Library Preparation) with sequencing on an Illumina Novaseq. The NGS data described in this validation document comes from 32 worksheets performed between January 2024 and August 2024.\

## Bioinformatic analysis\

The CLC pipeline, including details of the methodology and quality metrics, is described in the ERBB2 amplifications validation document (DOC6260). All amplification genes were analysed on the “coarse” setting, except for MYC. MYC has only 3 targets on the PanSolid enrichment, and preliminary work showed that the coarse setting could not identify MYC amplifications in positive controls. The “fine” setting was used specifically for MYC to account for the lower number of QIAseq targets. The version of the PanSolid CLC pipeline tested in this validation was PanSolid Workflow v2.2.

## Orthogonal testing\
### Droplet digital PCR for *ERBB2*, *MYC*, *MET*, *BRAF* and *EGFR*\

Amplification results for the ERBB2, MYC, MET, BRAF and EGFR genes were confirmed using droplet digital PCR (ddPCR). ddPCR assays were designed using the workflow for confirming constitutional CNVs, which involves a universal forward primer and a reference assay for the AP3B1 gene on chromosome 5 (DOC4304 CNV Confirmation Using Droplet Digital PCR). ddPCR was performed using a BioRad QX200 system with 5ng DNA input per well. Each assay was validated for use and then used to test DNA from clinical FFPE samples and SeraCare controls (Table 9). The results for the ERBB2 ddPCR were previously reported in DOC6260. Details of the primer and probe sequences, reaction mix (Table 19) and thermocycling conditions (Table 21) are given in the Appendix.\

### Whole genome sequencing\

Samples which had been tested using whole genome sequencing (WGS) were selected for this validation, which included samples from the Tessa Jowell BRAIN MATRIX Platform Study (NCT04274283) (13). WGS was performed on DNA extracted from fresh-frozen tissue by Illumina Laboratory Services at the Wellcome Trust Sanger Institute in Cambridge, as part of the NHS Genomics Medicine Service. Formalin-fixed paraffin-embedded (FFPE) tissue from the same tumour block as the fresh frozen tissue used for WGS was used as the sample source for PanSolid testing. Samples included in this validation were analysed by WGS between December 2022 to February 2024, using the “somatic variants and pertinent germline findings in cancer susceptibility genes” pipeline version 2.28 to version 3.6.2.\

## Validation analysis\

This validation document is written in Quarto using R (version 4.4.1) and is publicly available on [Github](https://github.com/joeshaw824/pansolid_cnv_validation).\

## Data locations\

All data analysed in this validation is saved here:\

{{< pagebreak >}}

# Results\
## Gene amplifications\

## Comparison to ddPCR\

```{r}
#| label: ddpcr-comparison
#| include: FALSE

ddpcr_cnv_amp_threshold <- 5

ddpcr_vs_pansolid <- validation_ddpcr_collated |> 
  mutate(sample_experiment = str_c(sample, "_", experiment)) |> 
  filter(target == "ch1_target" & 
           experiment %in% c("MYC Ex3_1",
                              "BRAF Ex10_1",
                              "MET Ex2_1",
                              "EGFR Ex6_1")) |> 
    filter(!duplicated(sample_experiment)) |>  
  inner_join(validation_all_amp_gene_results_collated,
             join_by(sample == labno, 
                     gene)) |> 
  mutate(labno_suffix_gene = str_c(labno_suffix, "_", gene)) |> 
  # Some samples were run twice on PanSolid
  filter(!duplicated(labno_suffix_gene)) |> 
  mutate(ddpcr_result = case_when(
    cnv >= ddpcr_cnv_amp_threshold ~"amplification",
    cnv < ddpcr_cnv_amp_threshold ~"normal result"),
    
    outcome = case_when(
      
      ddpcr_result == "amplification" & pansolid_call == "amplification" 
      ~"true_positive",
      ddpcr_result == "normal result" & pansolid_call == "normal result" 
      ~"true_negative",
      ddpcr_result == "amplification" & pansolid_call == "normal result"
      ~"false_negative",
      ddpcr_result == "normal result" & pansolid_call == "amplification" 
      ~"false_positive"
    ))

```

```{r}
#| label: tbl-ddpcr-vs-pansolid
#| tbl-cap: "PanSolid comparison to ddPCR"
#| echo: FALSE

ddpcr_summary <- ddpcr_vs_pansolid |> 
  count(gene, outcome) |> 
  pivot_wider(id_cols = gene,
              names_from = outcome,
              values_from = n,
              values_fill = 0) |> 
  mutate(`False positive` = 0) |> 
  rename(`Gene` = gene,
         `True negatives` = true_negative,
         `True positives` = true_positive,
         `False negatives` = false_negative)

kable(ddpcr_summary)

```

\

```{r}
#| label: tbl-ddpcr-false-negatives
#| tbl-cap: "False negative samples"
#| echo: FALSE

ddpcr_false_neg_table <- ddpcr_vs_pansolid |> 
  filter(outcome == "false_negative") |> 
  mutate(`PanSolid fold change` = round(max_region_fold_change, 1),
         `ddPCR copy number` = round(cnv, 1)) |> 
  rename(`ddPCR result` = ddpcr_result,
         `Lab number` = sample,
         `Gene` = gene,
         `PanSolid result` = pansolid_call) |> 
  select(`Lab number`,  `Gene`, `ddPCR copy number`, 
         `PanSolid fold change`, `ddPCR result`, `PanSolid result`)

kable(ddpcr_false_neg_table)

```

\

```{r}
#| label: fig-ddpcr-vs-pansolid-ngs
#| fig-cap: "ddPCR vs PanSolid by gene"
#| echo: FALSE

egfr_met_myc_plot <- ggplot(ddpcr_vs_pansolid |> 
         filter(gene != "MYC"), aes(x = cnv, y = max_region_fold_change)) +
  geom_point(shape = 21) +
  theme_bw() +
  labs(x = "ddPCR copy number",
       y = "PanSolid fold change",
       title = "ddPCR vs PanSolid by gene") +
  ggpubr::stat_cor(method = "pearson", label.x = 5, label.y = 0)  +
  facet_wrap(~gene)

myc_plot <- ggplot(ddpcr_vs_pansolid |> 
         filter(gene == "MYC"), aes(x = cnv, y = max_region_fold_change)) +
  geom_point(shape = 21) +
  theme_bw() +
  labs(x = "ddPCR copy number",
       y = "PanSolid fold change") +
  ggpubr::stat_cor(method = "pearson", label.x = 75, label.y = 10)  +
  facet_wrap(~gene)

combined_ddpcr_plot <- egfr_met_myc_plot + myc_plot + plot_layout(nrow = 2,
                                           heights = c(7,7))

combined_ddpcr_plot

```

{{< pagebreak >}}

## Comparison to WGS\

```{r}
#| label: select-wgs-samples
#| include: FALSE

pansolid_wgs_sample_selection <- validation_sample_patient_info |> 
  filter(nhsno %in% wgs_amp_data_reformatted_with_ids$nhs_no_clean) |> 
  filter(method_name == "QIAsymphony_DNA_FFPE") |> 
  # One patient has 2 QIAsymphony extractions with the same Pansolid results
  # I've arbitrarily chosen one to include
  filter(!labno %in% c("24030945", 
                      # This sample pathology block does not match the WGS
                      "24018922",
                      # This sample has a signal-adjusted noise above 1
                      "24030364")) |> 
  left_join(wgs_html_ids |> 
              select(patient_name, wgs_r_no, wgs_p_no, nhs_no_clean), 
            join_by("nhsno" == "nhs_no_clean")) 

```

```{r}
#| label: compare-wgs-and-pansolid
#| include: FALSE

wgs_pansolid_comparison <- validation_all_amp_gene_results_collated |> 
  filter(gene %in% amp_genes$gene) |> 
  filter(labno %in% pansolid_wgs_sample_selection$labno) |> 
  # One WGS sample was repeated for repeatability studies on worksheets
  # WS144291 and WS144291
  filter(!worksheet %in% c("WS144291", "WS144364")) |> 
  left_join(validation_sample_patient_info |> 
              select(labno, nhsno), by = "labno") |> 
  left_join(wgs_amp_data_reformatted_with_ids |> 
              select(nhs_no_clean, gene, wgs_result, wgs_pathno), 
            join_by(nhsno == nhs_no_clean, gene == gene)) |> 
  mutate(outcome = case_when(
      (wgs_result == "Gain detected" & pansolid_call == "amplification") ~"true_positive",
    
      (wgs_result == "No gain detected" & pansolid_call == "normal result") ~"true_negative",
    
      (wgs_result == "Gain detected" & pansolid_call == "normal result") ~"false_negative",
    
      (wgs_result == "No gain detected" & pansolid_call == "amplification") ~"false_positive"
    
  ))

```

```{r}
#| label: tbl-wgs-comparison
#| tbl-cap: "Comparison with WGS results"
#| echo: FALSE

wgs_comparison_tbl <- wgs_pansolid_comparison |> 
  group_by(gene, outcome) |> 
  count() |> 
  pivot_wider(id_cols = c(gene),
              names_from = outcome,
              values_from = n,
              values_fill = 0) |> 
  select(gene, true_negative, true_positive, false_negative) |> 
  mutate(`False positive` = 0) |> 
  rename(`Gene` = gene,
         `True negative` = true_negative,
         `True positive` = true_positive,
         `False negative` = false_negative)

kable(wgs_comparison_tbl)

```

{{< pagebreak >}}

## Limit of detection\

### SeraCare reference materials\

The Seraseq® TNA (DNA) WT Mix was tested on the PanSolid v2 enrichment on worksheet WS144291 and the +3 copies, +6 copies and +12 copies reference materials were tested on worksheet WS139890. The results for the 5 genes with defined copy number gains in the reference materials which are also on the list of targets for this validation are shown in  Table 1. The whole genome plot for each reference material is shown in Figure 3.\

The PanSolid pipeline results were consistently higher than the predicted fold change and the results provided by SeraCare for the EGFR and ERBB2 genes for the +3 copies, +6 copies and +12 copies reference materials. Conversely, the PanSolid results for MYC were consistently lower than the predicted fold chance and SeraCare results (Figure 5). This relationship was also observed when the ddPCR results for each reference material were compared to the ddPCR copy numbers provided by SeraCare (Figure 6).\

```{r}
#| label: seracare-ngs-comparison
#| include: FALSE

seracare_gene_copies <- read_csv(file = paste0(data_folder, 
                                                 "seracare_reference_materials/",
                                                 "seracare_gene_copies.csv")) |> 
  mutate(total_copies_ddpcr_seracare = additional_copies_ddpcr + 2,
         total_copies_ngs_seracare = additional_copies_ngs + 2,
         reference_material_short = factor(x = reference_material_short,
                                           levels = c("WT", "3 copies added",
                                           "6 copies added", "12 copies added")))

seracare_ids <- validation_all_amp_gene_results_collated |> 
  filter(!duplicated(labno)) |>
  select(labno, patient_name) |> 
  inner_join(seracare_gene_copies |> 
               filter(!duplicated(reference_material)) |> 
               select(reference_material),
             join_by("patient_name" == "reference_material"))


seracare_ngs_comparison <- validation_all_amp_gene_results_collated |> 
  # Remove limit of detection samples
  filter(!(worksheet == "WS144291" & labno ==  "24039973")) |> 
  # Remove samples run on PanSolid version 1
  filter(!worksheet == "WS138156") |> 
  inner_join(seracare_gene_copies, 
             join_by("patient_name" == "reference_material",
                     "gene" == "gene")) |> 
  mutate(
    total_copies_ngs_pansolid = max_region_fold_change * 2, 
    patient_name = factor(x = patient_name,
                               levels = c("CNVMix12CopiesSERASEQ",
                                          "CNVMix6CopiesSERASEQ",
                                          "CNVMix3copiesSERASEQ",
                                          "DNAWTmixSERASEQ")),
         
         predicted_fold_change = (expected_additional_copies + 2) / 2)

```

```{r}
#| label: tbl-seracare-pansolid-ngs
#| tbl-cap: "SeraCare reference material results"
#| echo: FALSE

seracare_ngs_comparison_tbl <- seracare_ngs_comparison |>
  mutate(`Pansolid fold change` = round(max_region_fold_change, 1),
         `PanSolid total copies` = round(total_copies_ngs_pansolid, 1),
         `SeraCare NGS total copies` = round(total_copies_ngs_seracare, 1)) |> 
  rename(`Lab number` = labno,
         `Reference material` = patient_name,
         `Gene` = gene,
         `Predicted fold change` = predicted_fold_change) |> 
  select(`Lab number`, `Reference material`, `Gene`, `Predicted fold change`, 
         `Pansolid fold change`,
         `PanSolid total copies`, `SeraCare NGS total copies`)

kable(seracare_ngs_comparison_tbl)

```

{{< pagebreak >}}

```{r}
#| label: fig-seracare-ngs-pansolid
#| fig-cap: "PanSolid vs SeraCare NGS"
#| echo: FALSE

seracare_colours <- c(
  "#009E73",
  "#F0E442",
  "#E69F00", 
  "#D55E00"
)

ngs_comparison_plot <- ggplot(seracare_ngs_comparison, 
                              aes(x = total_copies_ngs_pansolid,
                                    y = total_copies_ngs_seracare )) +
  geom_abline(linetype = "dashed") +
  geom_point(shape = 21, size = 3, alpha = 0.6, aes(fill = reference_material_short)) +
  scale_fill_manual(values = seracare_colours) +
  ggpubr::stat_cor(method = "pearson", label.x = 0, label.y = 30) +
  theme_bw() +
  theme(legend.position = "bottom") +
  scale_x_continuous(limits = c(-5, 35), 
                     breaks = c(0, 10, 20, 30)) +
  scale_y_continuous(limits = c(-5, 35), 
                    breaks = c(0, 10, 20, 30)) +
  facet_wrap(~gene) +
  labs(x = "Total copies NGS (Manchester)",
       y = "Total copies NGS (SeraCare)",
       fill = "")

ngs_comparison_plot

```

```{r}
#| label: seracare-ddpcr-comparison
#| include: FALSE

seracare_ddpcr_comparison <- validation_ddpcr_collated |> 
  filter(target_type == "Ch1Unknown") |> 
  inner_join(seracare_ids, join_by("sample" == "labno")) |> 
  inner_join(seracare_gene_copies, join_by("patient_name" == "reference_material",
                                            "gene" == "gene")) |> 
  mutate(patient_name = factor(x = patient_name,
                               levels = c("CNVMix12CopiesSERASEQ",
                                          "CNVMix6CopiesSERASEQ",
                                          "CNVMix3copiesSERASEQ",
                                          "DNAWTmixSERASEQ")))

```

```{r}
#| label: fig-seracare-ddpcr-pansolid
#| fig-cap: "ddPCR vs SeraCare"
#| echo: FALSE
#| warning: FALSE

ddpcr_comparison_plot <- ggplot(seracare_ddpcr_comparison, 
                              aes(x = cnv,
                                  y = total_copies_ddpcr_seracare )) +
  geom_abline(linetype = "dashed") +
  geom_point(shape = 21, size = 3, alpha = 0.6, aes(fill = reference_material_short)) +
  scale_fill_manual(values = seracare_colours) +
  facet_wrap(~gene) +
  ggpubr::stat_cor(method = "pearson", label.x = 0, label.y = 20) +
  theme_bw() +
  theme(legend.position = "bottom") +
  scale_x_continuous(limits = c(0, 25), 
                     breaks = c(0, 10, 20)) +
  scale_y_continuous(limits = c(0, 25), 
                     breaks = c(0, 10, 20)) +
  labs(x = "Total copies ddPCR (Manchester)",
       y = "Total copies ddPCR (SeraCare)",
       caption = "Pearson's coefficient not calculated for genes with only 2 data points",
       fill = "")

ddpcr_comparison_plot

```

{{< pagebreak >}}

### SeraCare dilution mixes\

Dilution mixes were prepared to simulate a sample with neoplastic cell contents of 30%, 20%, 10 % and 0%, and were tested on the PanSolid v2 enrichment on worksheet WS144291. The results for each gene are shown in Table 15, and the whole genome plots are shown in Figure 4. All gene amplifications were detected in the 30% NCC mix, except for MYC. \

```{r}
#| label: tbl-seracare-dilution-mix-results
#| tbl-cap: "SeraCare dilution mix results"
#| echo: FALSE

lod_dilution_mix_results <- validation_all_amp_gene_results_collated |> 
  filter(labno %in% c("24039973", "24039975") & worksheet == "WS144291") |> 
  filter(gene %in% c("ERBB2", "EGFR", "BRAF", "MET", "MYC")) |> 
  mutate(lod_percent = case_when(
    suffix == "a" ~30,
    suffix == "b" ~20,
    suffix == "c" ~10,
    suffix == "d" ~0
  ),
  lod_proportion = lod_percent/ 100,
  max_region_fold_change = round(max_region_fold_change, 1),
  predicted_fold_change = ((14 * lod_proportion) + (2 * (1 - lod_proportion))) / 2 ) |> 
  arrange(lod_percent) |> 
  rename(`Simulated NCC` = lod_percent,
         `Gene` = gene,
         `Predicted PanSolid fold change` = predicted_fold_change,
         `PanSolid fold change` = max_region_fold_change) |> 
  select(`Simulated NCC`, `Gene`, `Predicted PanSolid fold change`, `PanSolid fold change`)

kable(lod_dilution_mix_results)

```

{{< pagebreak >}}

## Repeat testing\

```{r}
#| label: fig-repeatability
#| fig-cap: "Repeatability results"
#| echo: FALSE

repeat_noise_data <- validation_stdev_results_collated |> 
  filter(duplicated(labno, fromLast = TRUE) |
           duplicated(labno, fromLast = FALSE)) |> 
  # Excel file produced twice as sample is in 2 cohorts
  filter(labno != "24030966") |> 
  # This is the limit of detection sample
  filter(labno != "24039973") |> 
  filter(!grepl(pattern = "SERA", x = patient_name,
               ignore.case = TRUE))

repeat_all_amp_data <- validation_all_amp_gene_results_collated |> 
  filter(labno %in% repeat_noise_data$labno)

low_fold_change_repeat_plot <- ggplot(repeat_all_amp_data |> 
         filter(labno != "24027774"), aes(x = gene, y = max_region_fold_change)) +
  geom_jitter(shape = 21, size = 2, width = 0.2) +
  theme_bw() +
  geom_hline(yintercept = 2.8, linetype = "dashed") +
  scale_y_continuous(limits = c(-2, 10),
                     breaks = c(-2, 0, 2.8, 5, 7.5, 10)) +
  theme(axis.text.x = element_text(angle = 90)) +
  facet_wrap(~labno) +
  labs(x = "", y = "Fold change")

high_fold_change_repeat_plot <- ggplot(repeat_all_amp_data |> 
         filter(labno == "24027774"), aes(x = gene, y = max_region_fold_change)) +
  geom_jitter(shape = 21, size = 2, width = 0.2) +
  theme_bw() +
  geom_hline(yintercept = 2.8, linetype = "dashed") +
  theme(axis.text.x = element_text(angle = 90)) +
  facet_wrap(~labno) +
  labs(x = "", y = "Fold change")

repeatability_plots <- low_fold_change_repeat_plot +
  high_fold_change_repeat_plot + plot_layout(nrow = 2,
                                             heights = c(5,5))

repeatability_plots

```

{{< pagebreak >}}

# Discussion\

# Review of test requirements\

# Requirement for clinical implementation\


# Conclusion\

{{< pagebreak >}}

# References\


{{< pagebreak >}}

# Appendix\

```{r}
#| label: tbl-ddpcr-primers
#| tbl-cap: "ddPCR primer sequences"
#| echo: FALSE
#| warning: FALSE

ddpcr_primers_probes <- read_csv(file = paste0(data_folder,
                                               "method_tables/",
                                               "ddpcr_primers.csv"))

ap3b1_primers <- c("Tailed_AP3B1_F", "AP3B1_R", "Universal F primer")

ddpcr_primer_tbl <- ddpcr_primers_probes |> 
  filter(category == "primer" & !name %in% ap3b1_primers) |> 
  select(-c(category, `5_label`, `3_label`)) |> 
  rename(`Primer name` = name,
         `Sequence` = sequence)

kable(ddpcr_primer_tbl)

```

```{r}
#| label: tbl-ap3b1-ddpcr-primers
#| tbl-cap: "AP3B1 universal reference primer sequences"
#| echo: FALSE
#| warning: FALSE

ap3b1_primers <- ddpcr_primers_probes |> 
  filter(category == "primer" & name %in% ap3b1_primers) |> 
  select(-c(category, `5_label`, `3_label`)) |> 
  rename(`Primer name` = name,
         `Sequence` = sequence)

kable(ap3b1_primers)

```

```{r}
#| label: tbl-ddpcr-probes
#| tbl-cap: "ddPCR probe sequences"
#| echo: FALSE
#| warning: FALSE

ddpcr_probe_tbl <- ddpcr_primers_probes |> 
  filter(category == "probe") |> 
  select(-category) |> 
  rename(`Primer name` = name,
         `Sequence` = sequence,
         `5 prime label` = `5_label`,
         `3 primer label` = `3_label`)

kable(ddpcr_probe_tbl)

```

```{r}
#| label: tbl-ddpcr-mix
#| tbl-cap: "ddPCR mix"
#| echo: FALSE
#| warning: FALSE

ddpcr_mix <- read_csv(file = paste0(data_folder, 
                                    "method_tables/",
                                    "ddpcr_mix.csv"))

kable(ddpcr_mix)

```

```{r}
#| label: tbl-ddpcr-universal-primer-mix
#| tbl-cap: "ddPCR universal primer mix"
#| echo: FALSE
#| warning: FALSE

ddpcr_universal_primer_mix <- read_csv(file = paste0(data_folder,
                                                     "method_tables/",
                                                     "ddpcr_universal_primer_mix.csv"))

kable(ddpcr_universal_primer_mix)

```

```{r}
#| label: tbl-ddpcr-thermocycler-programme
#| tbl-cap: "ddPCR thermocycler programme"
#| echo: FALSE
#| warning: FALSE

ddpcr_thermocycler_programme <- read_csv(file = paste0(data_folder,
                                                       "method_tables/",
                                                       "ddpcr_thermocycler_programme.csv"))

kable(ddpcr_thermocycler_programme)

```
