---
title: "PanSolid CNV Project Planning"
author: "Joe Shaw"
date: today
date-format: "DD/MM/YYYY"
format: pdf
editor: visual
---

```{r}
#| label: packages
#| include: FALSE

library(tidyverse)
library(readxl)
library(here)
library(epiR)
library(patchwork)

source(here::here("functions/cnv_functions.R"))

```

Here is the current plan for the different samples and variant types to be included in the project.

The idea is that the positive control samples for one variant type will be used as negative control samples for the other variant types.

LOH: loss of heterozygosity

```{r}
#| label: load-plan
#| echo: FALSE

variant_types <- read_excel(here("data/pansolid_cnv_variant_type_planning.xlsx")) 

total_samples <- sum(variant_types$positive_samples)

```

```{r}
#| label: show-plan-samples
#| echo: FALSE

variant_types_mod <- variant_types |> 
  mutate(negative_samples = total_samples - positive_samples) |> 
  select(-zygosity)

knitr::kable(variant_types_mod)

```

{{< pagebreak >}}

This is what the plan looks like in terms of the different variant categories.

```{r}
#| label: show-plan-categories
#| echo: FALSE

variant_categories <- variant_types |> 
  group_by(variant) |> 
  summarise(positive_samples = sum(positive_samples)) |> 
  mutate(negative_samples = total_samples - positive_samples)

knitr::kable(variant_categories)

```

```{r}
#| label: calculate-test-metrics
#| include: FALSE

get_calcs <- function(df, variant_type) {
  
  x <- df |> 
    filter(variant == variant_type)
  
  calcs <- summary(epiR::epi.tests(c(as.numeric(x[,2]), 
                             0, 0, as.numeric(x[,3])),
                           method = "clopper-pearson")) |> 
    mutate(type = variant_type)
  
  return(calcs)
  
}

```

If we assume that the results of the orthogonal method perfectly match the NGS PanSolid results (sensitivity and specificity = 100%), we can calculate the predicted test metrics for each variant type with 95% confidence intervals.

```{r}
#| label: perform-calcs
#| echo: FALSE

amps <- get_calcs(variant_categories, "Amplification")

dels <- get_calcs(variant_categories, "Whole gene deletion")

loh <- get_calcs(variant_categories, "Copy neutral LOH")

translocations <- get_calcs(variant_categories, "Translocation")

all_calcs <- rbind(amps, dels, loh, translocations)

all_calcs |> 
  filter(statistic %in% c("se", "sp")) |> 
  mutate(metric = case_when(
    
    statistic == "se" ~ "Sensitivity",
    
    statistic == "sp" ~ "Specificity"
    
  )) |> 
  ggplot(aes(x = type, y = est)) +
  geom_point(pch = 21, size = 2) +
  geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_wrap(~metric) +
  labs(x = "", y = "Proportion", 
       title = "Predicted test metrics for each variant type",
       caption = "Assuming 100% concordance with orthogonal test") +
  ylim(0.5, 1)

```

{{< pagebreak >}}

# Loss of Heterozygosity

Here's an example of how the loss of heterozygosity results could be presented.

```{r}
#| label: loh-analysis
#| include: FALSE
#| message: FALSE
#| warning: FALSE

example_loh <- read_xls(path = here::here("data/tLYNCH_WS130362_23013476.xls"),
         sheet = "SNV_calls") |> 
  rename(coordinate = start) |> 
  mutate(panel = "SNP") |> 
  janitor::clean_names()

example_loh_long <- example_loh |> 
  pivot_longer(cols = c("ref_percentage", "alt_percentage"),
               names_to = "column",
               values_to = "percentage") |> 
  mutate(allele = str_extract(column, pattern = "(\\w{3})_percentage",
         group = 1))

loh_chr3 <- example_loh_long |> 
               filter(chr == "chr3")

loh_chr2 <- example_loh_long |> 
  filter(chr == "chr2")

mlh1 <- all_transcripts |> 
  filter(transcript == "ENST00000231790") |> 
  mutate(gene = "")

msh2_msh6 <- all_transcripts |> 
  filter(transcript %in% c("ENST00000233146", "ENST00000234420")) |> 
  mutate(gene = "")

loh_mlh1_results <- ggplot(loh_chr3, aes(x = coordinate, y = percentage)) +
  geom_point(pch = 21, size = 3, alpha = 0.6, aes(fill = allele)) +
  scale_fill_manual(values = c(safe_blue, safe_red)) +
  theme_bw() +
  theme(legend.position = "bottom") +
  scale_y_continuous(limits = c(0, 100), breaks = seq(0, 100, by = 10)) +
  labs(x = "", y = "Allele percent (%)",
       title = "LOH results: chromosome 3",
       fill = "")

loh_msh2_6_results <- ggplot(loh_chr2, aes(x = coordinate, y = percentage)) +
  geom_point(pch = 21, size = 3, alpha = 0.6, aes(fill = allele)) +
  scale_fill_manual(values = c(safe_blue, safe_red)) +
  theme_bw() +
  theme(legend.position = "bottom") +
  scale_y_continuous(limits = c(0, 100), breaks = seq(0, 100, by = 10)) +
  labs(x = "", y = "Allele percent (%)",
       title = "LOH results: chromosome 2",
       fill = "")

mlh1_exon_plot <- ggplot(mlh1, aes(x = coordinate, y = gene)) +
  geom_segment(aes(x = start, xend = end, 
                     y = gene, yend = gene),
                 linewidth = 4) +
  geom_segment(aes(x = min(start), xend = max(end), 
                     y = gene, yend = gene),
                 linewidth = 0.5) +
  scale_x_continuous(limits = c(min(loh_chr3$coordinate), 
                                max(loh_chr3$coordinate))) +
  theme_bw() +
  labs(x = "Genomic coordinate (GRCh37)", y = "") +
  geom_label(aes(x = 37050000), label = "MLH1",
             nudge_y = -0.3)

msh2_6_exon_plot <- ggplot(msh2_msh6, aes(x = coordinate, y = gene)) +
  geom_segment(aes(x = start, xend = end, 
                     y = gene, yend = gene),
                 linewidth = 4) +
  scale_x_continuous(limits = c(min(loh_chr2$coordinate), 
                                max(loh_chr2$coordinate))) +
  theme_bw() +
  labs(x = "Genomic coordinate (GRCh37)", y = "") +
  geom_label(aes(x = 47650000), label = "MSH2",
             nudge_y = -0.3) +
   geom_label(aes(x = 48000000), label = "MSH6",
             nudge_y = -0.3)

loh_example_visual <- loh_msh2_6_results + loh_mlh1_results +
   msh2_6_exon_plot + mlh1_exon_plot +
  plot_layout(ncol = 2, nrow = 2, heights = c(4,2))

```

```{r}
#| label: loh-plot
#| echo: FALSE
#| fig-width: 10
#| fig-height: 10

loh_example_visual

```
