---
title: "PanSolid Copy Number Variant Limit of Detection Experiment"
author: "Joe Shaw"
date: today
date-format: "DD/MM/YYYY"
format: pdf
editor: visual
---

Code to produce this document is available on [GitHub](https://github.com/joeshaw824/pansolid_cnv_validation).

# Background

This is a limit of detection experiment using SeraCare reference materials to test the sensitivity of the PanSolid pipeline for detecting copy number variants at low neoplastic cell contents (NCC).

The aim of the experiment is to end up with 4 mixes, each representing low NCC samples (25%, 12.5%, 6.25% and 0% NCC) at the same total DNA input as a standard clinical sample:

-   20ul of 6ng/ul DNA
-   120ng DNA
-   36,363 haploid genome copies (1 haploid human genome weighs 3.3 picograms)

Of this, the following amounts are taken as the input into the fragmentation, end-repair and A-addition reaction:

-   16.75ul
-   100.5ng
-   30,454 haploid genome copies

{{< pagebreak >}}

## Samples

The "tumour" sample is the Seraseq® Solid Tumor CNV Mix (+12 copies of various genes including ERBB2).

The "normal" sample is the Seraseq® DNA WT Mix (no additional gene copies).

```{r}
#| label: sample-volumes
#| include: TRUE
#| message: FALSE
#| warning: FALSE

library(tidyverse)

tumour_stock_vol <- 15.8

tumour_stock_conc <- 10.9 

norm_stock_vol <- 14.9

norm_stock_conc <- 28.8

tumour_total_ng <- tumour_stock_vol * tumour_stock_conc

norm_total_ng <- norm_stock_vol * norm_stock_conc

```

{{< pagebreak >}}

# Calculations

```{r}
#| label: calculations
#| include: TRUE
#| message: FALSE

total_c2 <- 6

total_v2 <- 20

total_ng <- total_c2 * total_v2

tumour_additional_copies <- 12

df <- data.frame("target_tumour_percent" = c(25, 12.5, 6.25, 0),
                 "tumour_c1" = c(tumour_stock_conc / 2, tumour_stock_conc / 4, 
                                 tumour_stock_conc / 8, 0),
                 "norm_c1" = c(norm_stock_conc/2, norm_stock_conc/2, norm_stock_conc/2, 
                               norm_stock_conc/2))

lod_calculations <- df |> 
  mutate(
    
    target_norm_percent = 100-target_tumour_percent,
    
    # Calculate nanograms required of each sample (normal and tumour)
    
    tumour_ng = round(total_ng * (target_tumour_percent / 100), 1),
    
    norm_ng = round(total_ng * (target_norm_percent / 100), 1),
    
    total_ng = tumour_ng + norm_ng,
    
    # Calculate number of genome equivalents (GE) 
    # 1 haploid human genome is 3.3 picograms
    
    tumour_ge = round((tumour_ng * 1000) / 3.3, 1),
    
    norm_ge = round((norm_ng * 1000) / 3.3, 1),
    
    total_ge = tumour_ge + norm_ge,
    
    # Calculate the starting volumes (v1) of each sample
    
    tumour_v1 = ifelse(tumour_ng ==0, 0, round(tumour_ng / tumour_c1, 1)),
    
    norm_v1 = ifelse(norm_ng == 0, 0, round(norm_ng / norm_c1, 1)),
    
    # Calculate the water volume required
    
    water_v1 = total_v2 - (tumour_v1 + norm_v1),
    
    final_v2 = water_v1 + tumour_v1 + norm_v1,
    
    final_c2 = round(((tumour_v1*tumour_c1) + (norm_v1*norm_c1)) / final_v2,0),
    
    # Calculate the additional number of copies in each mix
    
    norm_copies = round((target_norm_percent / 100) * 2,2),
    
    tumour_copies = round((target_tumour_percent / 100) * (tumour_additional_copies + 2),2),
    
    total_copies = round((norm_copies + tumour_copies),2),
    
    total_additional_copies = total_copies - 2,
    
    expected_fold_change = total_copies / 2
    
    )

# Check that we have enough of each control

sum(lod_calculations$tumour_ng) < tumour_total_ng

sum(lod_calculations$norm_ng) < norm_total_ng

lod_table <- lod_calculations |> 
  select(target_tumour_percent, tumour_c1, 
         norm_c1, tumour_v1, norm_v1, water_v1, 
         final_v2, final_c2) |> 
  rename("Target tumour %" = target_tumour_percent)
  
knitr::kable(lod_table)

```

```{r}
#| label: export-results
#| include: FALSE

write.csv(lod_calculations, file = here::here("outputs/lod_calculations.csv"),
          row.names = FALSE)

```

{{< pagebreak >}}

# Lab Work

I set up these dilutions on 01/03/2024 and they were run on worksheet WS139681.

I used the Qubit values from the Seraseq technical product report -- there was such limited quantities that I didn't want to use up some of the sample for additional Qubit.

My initial plan was to Qubit the serial dilutions, as Eleanor had suggested. I wanted to use a High Sensitivity Qubit but there aren't any in pre-PCR, so I used the Broad Range which should have been sensitive down to 0.1ng/ul.

But the Qubit values for each of the serial dilutions were "out of range". So I continued with the set up and checked the final concentration of each mix using the Broad Range Qubit.

{{< pagebreak >}}

![Limit of Detection Experiment](C:/Users/joseph.shaw2/Documents/pansolid_cnv_validation/images/joseph.shaw2_04-03-2024_09-26-29-PR-24977-RICOH.pdf)
