---
title: "PanSolid CNV Analysis"
author: "Joe Shaw"
format: pdf
editor: visual
---

## Introduction

This Quarto document is a notebook for organising analysis of somatic CNVs from a CLC Genomics Workbench pipeline (Qiagen).

## Reading in the Data

```{r}

library(tidyverse)
library(readxl)

datapath <- "S:/central shared/Genetics/NGS/Bioinformatics/1_Pan-solid-Cancer/CNV/00_Amplifications_Fine_vs_Coarse/"

cns_files <- list.files(str_c(datapath, "CNS_PS/"), full.names = TRUE)

list.f

filename_regex <- regex(
  r"[
  (WS\d{6})             # Worksheet number
  _
  (\d{8})               # Sample number
  _
  ([A-z]+)              # Patient name
  .xlsx
  ]",
  comments = TRUE
)


get_gene_result <- function(df, gene_name) {
  
  x <- df |> 
    filter(gene == gene_name)
  
  output <- case_when(
    nrow(x) < 1 ~"No call",
    nrow(x) >= 1 ~"Amplification")
  
  return(output)
  
}

summarise_results <- function(file, input_sheet) {
  
  results <- read_excel(path = file,
                        sheet = input_sheet) |> 
    janitor::clean_names()
  
  sample_id <- str_extract(file, filename_regex,
                           group = 2)
  
  patient_name <- str_extract(file, filename_regex,
                           group = 3)
  
  egfr_result <- get_gene_result(df = results, gene_name = "EGFR")
  
  erbb2_result <- get_gene_result(df = results, gene_name = "ERBB2")

  met_result <- get_gene_result(df = results, gene_name = "METT")

  summary <- tribble(
  
  ~sample, ~name, ~gene, ~result,
  sample_id, patient_name, "EGFR", egfr_result,
  sample_id, patient_name, "ERBB2", erbb2_result,
  sample_id, patient_name, "MET", met_result
  )
  
  return(summary)
  
}

collated <- cns_files |>
  map(\(cns_files) summarise_results(
    file = cns_files,
    input_sheet = "Oncogenes (Amplified) Coars..."
  )) |>
  list_rbind()

```
